added solve_sylvester and accompanying tests

Dekermanjian opened the pull request:
Hello, JAX team!

I am new to working with JAX and would appreciate feedback (and guidance) on my implementation of a solver for the Sylvester equation with accompanying tests. 

This is related to #6089  #669  and would be pre-requisite to #19109 

Thanks in advance!

Dekermanjian committed:
added solve_sylvester and accompanying tests

1. moved solve_sylvester and associated tests to appropriate locations in codebase
2. Updated docstring to conform to codebase conventions
3. Updated solve_sylvester to return matrix of NaNs when no solution is available
4. Updated tests to reflect changes in solve_sylvester as well as utilize parameterization routines for testing different shapes and dtypes where appropriate

cleaned up imports that I carelessly left behind

1. updated tests to use _CheckAgainstNumpy and _CompileAndCheck
2. updated tests to NOT change JAX configurations
3. Moved tests into ScipyLinalgTest
4. updated solve_sylvester to avoid intermediate arrays of different types
5. updated solve_sylvester to check for incoming shapes and raise ValueError on incorrect shapes
6. Added disclaimer to docstring on float32 precision issues

1. added solve_sylvester to docs
2. updated casting of matrix to specific type
3. updated function type hints
4. updated tests

1.updated solve_sylvester to work with complex types
2. updated tests to test complex types
3. updated utility function for random invertible matrices
4. updated docstring for no sylvester solution test

resolved merged conflicts

1. updated docstring for lower precision printing
2.specified cpu and gpu only for the tests

removed dtype tolerance overides for float64 and complex128

removed precision float32 warning from docstring

1. reworked sylvester solver to use the Bartel-Stewart algorithm
2. Added a tol argument to allow the user to specify how degenerate the A and B matrices can be and still return a solution

1. added both Bartel-Stewart and eigen decomposition algorithms
2. Updated the docstring with more information about each method
3. Added a tol argument to the eigen decompositiom algorithm

consolidated eigen and schur sylvester methods into one with a method argument

1. updated docstring to include method arg
2. removed jnp.array from inside promote_args_inexact
3. skip GPU for schur method in tests

fixed missing import after conflict resolution and changed static_argnum to static_argnames

made method and tol keyword only arguments

changed Literal typing to string. updated ifelse for method in solve sylvester and added raise valueerror. fixed spacing formatting to two space indentations

updated isreal type checking and return lax.cond instead of jax.lax.cond

squashed commits into one

jakevdp commented on jax/_src/third_party/scipy/linalg.py:
```diff
@@ -110,3 +112,74 @@ def funm(A: ArrayLike, func: Callable[[Array], Array],
+        )
+
+    # Check if there is a solution first
+    is_sylvester_solvable(A, B)
```
We should not do this kind of runtime value checking in JAX library routines. There is no supported API for raising value-dependent runtime errors (and raising runtime errors in callbacks has issues of its own)

Dekermanjian replied:
@jakevdp, my apologies for this type of implementation. I am still getting my feet wet with JAX and wasn't sure how to handle this with JAX. Is it more appropriate to handle this by returning a matrix of NaNs if there is no solution?

jakevdp replied:
Yes, returning NaNs or infs would be appropriate here.

jakevdp commented on tests/third_party/scipy/solve_sylvester_test.py:
```diff
@@ -0,0 +1,34 @@
+import pytest
```
`tests/third_party/scipy` should only be used for tests substantially copied from other packages, and in that case should include a license and attribution.

I'm not sure whether this code was copied from elsewhere or if the location of the test was inadvertent, but in either case it would be better to write a new test method from scratch and put it in `tests/linalg_test.py`. Please use the test parameterization routines there to test for a variety of input shapes and dtypes.

Dekermanjian replied:
Thank you for the clarification. I wrote these tests from scratch and will move them to the appropriate location and will adhere to your suggestion of using the test parameterization routines.

jakevdp commented on jax/_src/third_party/scipy/linalg.py:
```diff
@@ -110,3 +112,74 @@ def funm(A: ArrayLike, func: Callable[[Array], Array],
+def solve_sylvester(A: jnp.ndarray, B: jnp.ndarray, C: jnp.ndarray) -> jnp.ndarray:
+    """
+    Solves the Sylvester equation:
+      .. math:: AX + XB = C
```
This is incorrectly formatted RST. Either use a proper `.. math::` directive, or use an inline `:math:` role. More info on formatting at https://sphinx-rtd-trial.readthedocs.io/en/latest/ext/math.html

Dekermanjian replied:
Thank you for catching that! I will fix this in accordance with the resource that you've linked.

jakevdp commented on jax/_src/third_party/scipy/linalg.py:
```diff
@@ -110,3 +112,74 @@ def funm(A: ArrayLike, func: Callable[[Array], Array],
+    Solves the Sylvester equation:
+      .. math:: AX + XB = C
+
+    Parameters
```
Please use the docstring style that matches the rest of the file (`Args:` rather than `Parameters`, no underline on sections, parameter descriptions without types on their own line, etc.)

jakevdp commented on tests/third_party/scipy/solve_sylvester_test.py:
```diff
@@ -0,0 +1,34 @@
+import jax
+from jax import grad, random
+import jax.numpy as jnp
+from jax._src.third_party.scipy.linalg import solve_sylvester
```
Tests should import from public rather than private paths (`jax.scipy.linalg.solve_sylvester`)

Dekermanjian replied:
Thank you for this clarification. I will make the suggested change when importing the `solve_sylvester()` function

jakevdp commented on jax/_src/third_party/scipy/linalg.py:
```diff
@@ -2,11 +2,13 @@
 
 from collections.abc import Callable
 
+import jax
```
`third_party/scipy/linalg.py` is only for code copied substantially from third party sources, and if so it should include an attribution.

I'm not sure the source of this code, but we should prefer to rewrite from scratch rather than copying existing code, regardless of the license.

Dekermanjian replied:
Thank you for the clarification, I will move the code to the appropriate location.

jakevdp reviewed (changes requested):
Thanks for the contribution! In order to land this in JAX it will have to be substantially modified – please start by addressing the comments below, and then we can iterate from there. Thanks!

Dekermanjian commented:
Thank you, @jakevdp for your constructive feedback! I will make changes and modify the implementation in accordance with your suggestions.

Dekermanjian commented:
Hi @jakevdp, I made the changes that you suggested above. I have a quick question, in the tests after inheriting from `jtu.JaxTestCase` it seems like the `numpy_dtype_promotion` is set to `strict`. This was  causing my tests to fail. Inside my tests, I set `numpy_dtype_promotion=standard` which is most likely incorrect. Could you point me to some documentation about this issue so that I can fix it?

jakevdp commented on tests/linalg_test.py:
```diff
@@ -2357,5 +2358,59 @@ def testPascal(self, n, kind):
     self._CompileAndCheck(jsp_fun, args_maker)
 
 
+class TestSolveSylvester(jtu.JaxTestCase):
```
No need to make a new test class for this single function; instead let's add new test cases to the existing `ScipyLinagTest`.

jakevdp commented on tests/linalg_test.py:
```diff
@@ -2357,5 +2358,59 @@ def testPascal(self, n, kind):
 
+class TestSolveSylvester(jtu.JaxTestCase):
+  # float32 is not close enough for assert
+  jax.config.update("jax_enable_x64", True)  # Ensure float64 for higher precision
```
Please do not update configurations as part of the test. Every test file is run in both 32 and 64-bit precision; if 32-bit precision is problematic for some reason, we can skip those tests (and note this in the function documentation)

jakevdp commented on tests/linalg_test.py:
```diff
@@ -2357,5 +2358,59 @@ def testPascal(self, n, kind):
+    X_true = random.normal(random.key(2), shape=(n, m))
+
+    C = A @ X_true + X_true @ B
+    with jax.numpy_dtype_promotion('standard'):
```
If all three inputs are the same dtype, then the function should execute without error under strict dtype promotion. If it does not, it should be fixed within the `sylvester` implementation: it means you're generating intermediate arrays of mixed type, and instead you should explicitly cast them within array operations.

jakevdp commented on tests/linalg_test.py:
```diff
@@ -2357,5 +2358,59 @@ def testPascal(self, n, kind):
+    # Test no solution case to AX + XB = C
+    # For example, the simple case A=[[1]], B = [[-1]], and C = [[1]] would generate (1)X + X(-1) = (1)
+    # which equals X - X = (1) which is not possible because X - X = 0
+    with jax.numpy_dtype_promotion('standard'):
```
Same here: we should remove the standard dtype promotion. Everything in this block should execute fine with strict promotion.

jakevdp reviewed (changes requested):
It's looking better! A few additional comments below.

jakevdp commented on jax/_src/scipy/linalg.py:
```diff
@@ -2243,3 +2243,45 @@ def _binom(n, k):
+
+
+@jit
+def solve_sylvester(A: jnp.ndarray, B: jnp.ndarray, C: jnp.ndarray) -> jnp.ndarray:
```
Input arrays should be annotated with `ArrayLike`, and output should be annotated with `Array`, similar to other APIs in this file.

With this change, you'll have to preprocess the inputs (using e.g. `promote_args_inexact`) to convert them to JAX arrays using `jnp.asarray`, or else you'll get type checking errors because `ArrayLike` doesn't support the `.shape` attribute

jakevdp reviewed (changes requested):
Looking great, we're getting closer!

Also, to ensure this appears in docs, we'll need to add this new API to the list here: https://github.com/jax-ml/jax/blob/1d10a488d524b3ad1562f3ef0c0c74030913a5f9/docs/jax.scipy.rst?plain=1#L77

Dekermanjian commented on jax/_src/scipy/linalg.py:
```diff
@@ -2243,3 +2243,45 @@ def _binom(n, k):
+
+    RA, UA = jnp.linalg.eig(A)
+    RB, UB = jnp.linalg.eig(B)
+    F = solve(UA, C.astype(RA.dtype) @ UB)
```
Here the issue is that `jnp.linalg.eig()` is changing the dtype from a float type to a complex type. I changed the casting to explicitly be `C.astype("complex")`. Let me know if that is still not correct.

Dekermanjian commented on tests/linalg_test.py:
```diff
@@ -2114,6 +2115,56 @@ def testToeplitzConstructionWithKnownCases(self):
+
+    C = rng(shape=(n, n), dtype=dtype)
+    sylv_solution = solve_sylvester(A, B, C)
+    assert jnp.isnan(sylv_solution).all()
```
Hey good call on checking this. Turns out SciPy had a bug with this exact issue. So I opened up and issue and will fix the bug over at SciPy according to their preferences. https://github.com/scipy/scipy/issues/23066

jakevdp replied:
OK, in that case we can skip the scipy check (even if the upstream bug is fixed, we'll have to test against older scipy versions for some time still)

For the assertion, let's use `self.assertArraysEqual`, becuase it gives a more informative error in case of a test failure:
```python
self.assertArraysEqual(sylv_solution, np.full((n, n), np.nan, dtype))
```

Dekermanjian commented:
Hey @jakevdp, thank you for your patience with this PR. I made most of the changes you suggested above. I left replies on the suggestions where I took a different approach above. Let me know if my different approach is incorrect and I will go back change it.

jakevdp commented on tests/linalg_test.py:
```diff
@@ -68,6 +68,21 @@ def _axis_for_ndim(ndim: int) -> Iterator[None | int | tuple[int, ...]]:
     yield (-1, 0, 1)
 
 
+def _random_invertible(rng, shape, dtype):
+    """
+    Generate a random invertible matrix was specified shape and dtype
+    """
+    invertible = False
+    while not invertible:
+      a = rng(shape, dtype)
+      try:
+        np.linalg.inv(a)
+        invertible = True
+      except np.linalg.LinAlgError:
+        pass
+    return [a]
```
Suggestion for a slightly more compact definition – also, please replace the other use of this code with the helper function.

(note in particular: I think `return a` rather than `return [a]` makes more sense in this context)
```suggestion
def _random_invertible(rng, shape, dtype):
    """Generate a random invertible matrix was specified shape and dtype"""
    while True:
      a = rng(shape, dtype)
      try:
        np.linalg.inv(a)
      except np.linalg.LinAlgError:
        pass
      else:
        return a
```

jakevdp commented on tests/linalg_test.py:
```diff
@@ -2114,6 +2129,54 @@ def testToeplitzConstructionWithKnownCases(self):
 
+  @jtu.sample_product(
+    shape=[(2, 3), (4, 6), (5, 7), (100, 300)],
+    dtype = float_types
```
Should we test complex types as well?

jakevdp commented on tests/linalg_test.py:
```diff
@@ -2114,6 +2129,54 @@ def testToeplitzConstructionWithKnownCases(self):
+  )
+  def test_no_solution_sylvester(self, n, dtype):
+    # Test no solution case to AX + XB = C
+    # For example, the simple case A=[[1]], B = [[-1]], and C = [[1]] would generate (1)X + X(-1) = (1)
```
It seems this example doesn't match the code below – can we change this to a description of the condition being used in the code?

jakevdp commented on tests/linalg_test.py:
```diff
@@ -2114,6 +2129,54 @@ def testToeplitzConstructionWithKnownCases(self):
+    # Define eigenvalues that sum to zero
+    eigenvalues_A = rng(shape=(n,), dtype=dtype)
+    eigenvalues_B = -eigenvalues_A
+    P = _random_invertible(rng=rng, shape=(n, n), dtype=dtype)[0]
```
Remove the `[0]` here once you change the function to `return a` rather than `return [a]`

Dekermanjian commented:
Hey @jakevdp, I made the above changes and modified `solve_sylvester()` to output in the dtype that matches the input dtype. Allowing for complex or float outputs.

jakevdp commented on tests/linalg_test.py:
```diff
@@ -68,6 +68,21 @@ def _axis_for_ndim(ndim: int) -> Iterator[None | int | tuple[int, ...]]:
+      a = rng(shape, dtype)
+      try:
+        np.linalg.inv(a)
+        invertible = True
```
remove `invertible = True`; it's unused.

jakevdp commented on jax/_src/scipy/linalg.py:
```diff
@@ -2243,3 +2243,51 @@ def _binom(n, k):
+      >>> B = jax.numpy.array([[5, 6], [7, 8]])
+      >>> C = jax.numpy.array([[6, 8], [10, 12]])
+      >>> jax.scipy.linalg.solve_sylvester(A, B, C)
+      Array([[1.00000000e+00, 2.36169983e-16],
```
This doctest is failing, because the exact rounding of values near zero differs between platforms. Typically we address this by adjusting the print options; for example: https://github.com/jax-ml/jax/blob/2c838d4ae3fff34242d5a2993e6892c8214c9177/jax/_src/scipy/linalg.py#L805-L813

jakevdp commented on tests/linalg_test.py:
```diff
@@ -2114,6 +2122,58 @@ def testToeplitzConstructionWithKnownCases(self):
+    shape=[(2, 3), (4, 6), (5, 7), (100, 300)],
+    dtype = float_types + complex_types
+  )
+  def test_solve_sylvester(self, shape, dtype):
```
This test is failing on TPU because non-symmetric `eig` is not implemented. You can fix this by specifying that the test should only run on CPU and GPU; see the `eig` tests for an example of this: https://github.com/jax-ml/jax/blob/da106b971af143906c8cbec8f4f2dfb7b57a11ce/tests/linalg_test.py#L268-L269

Dekermanjian commented on tests/linalg_test.py:
```diff
@@ -2114,6 +2122,58 @@ def testToeplitzConstructionWithKnownCases(self):
+      X_true = rng(shape=(m, n), dtype=dtype)
+
+      C = A @ X_true + X_true @ B
+      print(C.dtype)
```
I apologize for the above, these were careless of me to miss.

jakevdp commented on jax/_src/scipy/linalg.py:
```diff
@@ -2243,3 +2243,53 @@ def _binom(n, k):
+
+    Notes:
+      This function returns NaNs in the event that the eigenvalues of the A and B matrices sum to zero elementwise.
+      IMPORTANT: Float32 precision may not be adequate, especially with larger matrices, to match expected results consider float64 precision.
```
I'm trying to understand this warning: what is the precise problem with float32 precision?

Dekermanjian replied:
@jakevdp  No, my intention was only to adjust the tolerance on float32. I set the tolerance on float64 to 1e-5 to match what `jnp.allclose()` says is the default tolerance `rtol`.  Is the tol that is passed to `_CheckNumpy` adjusting the `atol` instead?

Dekermanjian replied:
Yes, that makes sense. I will go ahead and remove overrides to float64 and complex128. 

In practice, this means that your estimated $\hat{X}$ from AX + XB = C will be off from the true X by more than the default tolerance value of `jnp.allclose()`. This difference between $\hat{X}$ and $X$ may not be acceptable depending on the required precision. I will also mention that the difference between $\hat{X}$ and $X$ is larger than the difference between $\hat{X}$ and the estimated $\tilde{X}$ returned from scipy.'s `solve_sylvester`.

jakevdp commented:
On h100 the large test on float32 is failing with an even larger discrepancy:
```
Mismatched elements: 8895 / 30000 (29.6%)
Max absolute difference among violations: 1.5408611
Max relative difference among violations: 18452.156
 ACTUAL: array([[ 1.800829,  2.928643, -0.344496, ...,  4.774832, -2.07168 ,
        -5.008069],
       [ 1.854761, -3.498744, -1.515197, ...,  0.164751,  2.064151,...
 DESIRED: array([[ 1.785693,  2.994908, -0.187475, ...,  4.250907, -2.186043,
        -4.966815],
       [ 1.842348, -2.990138, -0.931781, ...,  0.104629,  2.06976 ,...
```
I'm not sure the best approach here – is there some fundamental issue with the algorithm used here?

Dekermanjian commented:
> On h100 the large test on float32 is failing with an even larger discrepancy:
> 
> ```
> Mismatched elements: 8895 / 30000 (29.6%)
> Max absolute difference among violations: 1.5408611
> Max relative difference among violations: 18452.156
>  ACTUAL: array([[ 1.800829,  2.928643, -0.344496, ...,  4.774832, -2.07168 ,
>         -5.008069],
>        [ 1.854761, -3.498744, -1.515197, ...,  0.164751,  2.064151,...
>  DESIRED: array([[ 1.785693,  2.994908, -0.187475, ...,  4.250907, -2.186043,
>         -4.966815],
>        [ 1.842348, -2.990138, -0.931781, ...,  0.104629,  2.06976 ,...
> ```
> 
> I'm not sure the best approach here – is there some fundamental issue with the algorithm used here?

Hey @jakevdp, there is a difference between the algorithm here and the one that scipy uses. SciPy first does a Schur decomposition of the matrices and solves a simpler solution before recasting the matrices back to the original basis. I had a working implementation of this but it was a little bit slow because of some nested for loops. 

Would you give me a couple of days to try and take that solution and implement it using JAX scans? I think that it will be worth it for the algorithm to be more robust.

Dekermanjian commented:
Hey @jakevdp, I was able to implement the same algorithm that SciPy uses to solve the sylvester equation (Bartel-Stewart algorithm). 

I believe that it now produces more similar results to SciPy (when you force scipy to use float32 instead of casting to float64). However, in my testing I found that when the A and B matrices have eigenvalues that sum to values close to zero the results become poor, especially for float32. With float64 depending on how close to zero the sums are you may still get pretty good approximations (the above also applies to SciPy) and because of that I added in a `tol` argument in the `solve_sylvester` function that will allow the user to specify how close to zeros the sum of the eigenvalues can be before returning a matrix of NaNs. 

Now because this affects float32 more drastically than float64, I wonder if it warrants adding back the warning to the docstring?

jakevdp commented:
> Would you be opposed to having both methods available and leaving it up to the user to select which one they would rather use?

That would be a great solution I think. We'd have to make sure to provide enough documentation to let the user make an informed choice of the setting.

Dekermanjian commented:
Okay, I will go ahead and bring back the eigenvalue decomposition method and I will beef up the docstrings to reflect what each method is doing behind the scenes.

Dekermanjian commented:
Hey @jakevdp, I am experiencing something weird when testing. I am getting unexpected results only during tests.

Below you can see the results of `pytest -n auto tests/linalg_test.py -k test_no_solution_sylvester_eigen`:

<img width="1268" alt="Screenshot 2025-06-03 at 4 46 16 PM" src="https://github.com/user-attachments/assets/d27ca922-776d-4048-b09f-d9e8309043c8" />

So I take the generated A, B, and C matrices and run them through the exact same function outside of the testing environment and this results in what I expect it to result in; a matrix of NaNs.

<img width="1080" alt="Screenshot 2025-06-03 at 4 47 34 PM" src="https://github.com/user-attachments/assets/62c81f7d-a22a-4925-99a0-0ecc95e0a885" />

Am I missing something obvious here?

jakevdp commented:
Are you reconstructing the matrices from their printed representation? That may not result in bitwise-identical floating point values.

Dekermanjian commented:
Okay, @jakevdp I saved the matrices to disk using `np.save()` and loaded them outside the testing environment with `np.load()` and as expected the solver returns a matrix of NaNs only outside the testing environment. However during the test I am getting odd results. Any idea what might be happening?

jakevdp commented:
Is it possible that it's a float64 precision issue? Do you have the same X64 flag set in both contexts?

Dekermanjian commented:
Yeah, I believe I do. I haven't enabled float64 in both settings. The matrices of the failing tests are when testing complex64 and I checked the dtype of the matrices after `np.load()` and can confirm that they were loaded with complex64 dtype.

Dekermanjian commented:
Hey I might have found the issue. On the other environment I am running JAX version 0.6.1. While on the testing environment I am running JAX 0.6.1.dev20250602+f067b6586. I can confirm that on that JAX version I am not getting the expected results whereas on 0.6.1 I am. Any idea why that could be?

Dekermanjian commented:
Okay, @jakevdp after digging into the two versions of JAX I found that the results of `jnp.linalg.eig()` in the dev version is returning different results to JAX 0.6.1 and that the results from JAX 0.6.1 compared to the results from `np.linalg.eig()` match more closely than those returned by the JAX dev version. Moreover, since the matrices A, B and C are simulated to have sum zero eigenvalues the results from JAX 0.6.1 and numpy are correct as they recover this property while the dev version of JAX does not. 

I tried looking at recent commits that may have changed `jnp.linalg.eig()` but I could not really pin point where the issue arises. Any ideas here?

jakevdp commented:
I don't think `eig` has changed between 0.6.1 and the current version. I suspect that somehow the jaxlib you're using in each environment is targeting a different LAPACK under the hood, and this is leading to different numerics between the environments.

Dekermanjian commented:
@jakevdp I checked the version on `jaxlib` and in my dev environment I was running `jaxlib=0.6.0` so I updated it to version 0.6.1 to match the environment that is returning the expected `jnp.linalg.eig()` results but it is still returning incorrect values. Is there a way for me to verify what LAPACK version is being called by `jnp.linalg.eig()`?

Dekermanjian commented:
Hi @jakevdp, is there anything you'd like me to add/remove/modify with the latest commit?

jakevdp commented:
Hi - it wasn't clear to me whether you'd solved the `eig` issue. Is that resolved now?

Dekermanjian commented:
Sorry, about that! I should have communicated better. In the last commit I brought back the Eigen decomposition method as another option with a tolerance argument for letting the user decide how to handle ill-conditioned matrices. I added to the docstrings of both methods (Eigen and Bartel-Stewart) descriptions and the strengths and weaknesses of each method. 

For the issue with the Eigen where different environments are returning different results. One environment is returning the same results as `numpy.linalg` while the other is the same as `scipy.linalg`, the differences arise due to calling different LAPACK's and the differences are specific to low-precision and ill-conditioned matrices. All implementations (numpy, scipy, and our JAX implementations) struggle with low-precision + ill-conditioned matrices. 

I think that maybe it would be better to re-purpose the second test `test_no_solution` to `test_tol` which will focus on testing the tolerance argument against ill-conditioned matrices. 

Does this sound like a viable path forward?

jakevdp commented on docs/jax.scipy.rst:
```diff
@@ -75,6 +75,8 @@ jax.scipy.linalg
    schur
    solve
+   solve_sylvester_eigen
+   solve_sylvester_schur
```
Rather than having different functions, I think it would be better to have a single `solve_sylvester` function with a `method` parameter that lets you choose `method='eigen'` or `method='schur'`.

jakevdp reviewed (commented):
Looks great! Just some small comments left

Dekermanjian commented:
Thanks again for your patience @jakevdp. I made the suggested changes and re-pushed.

jakevdp reviewed (commented):
Looks good! One more minor comment, and I'll trigger the tests

Dekermanjian commented on jax/_src/scipy/linalg.py:
```diff
@@ -2248,3 +2248,153 @@ def _binom(n, k):
+    Returns:
+      Y: Matrix of shape m x n
+    """
+    R, S, F = promote_args_inexact("_solve_sylvester_triangular_scan", R, S, F)
```
I am sorry about this. I must have missed this after resolving the merge conflicts.

Dekermanjian commented:
I swapped out `static_argnums` for `static_argnames` and fixed the missing import issue. Please let me know if there is anything else I can do!

jakevdp commented on jax/_src/scipy/linalg.py:
```diff
@@ -2248,3 +2248,153 @@ def _binom(n, k):
+
+
+@partial(jit, static_argnums=[3, 4])
+def solve_sylvester(A: ArrayLike, B: ArrayLike, C: ArrayLike, method: Literal["eigen", "schur"] = "schur", tol: float = 1e-8) -> Array:
```
To make them keyword-only arguments, it should look like this:
```suggestion
def solve_sylvester(A: ArrayLike, B: ArrayLike, C: ArrayLike, *, method: Literal["eigen", "schur"] = "schur", tol: float = 1e-8) -> Array:
```

Dekermanjian replied:
Actually, would you mind if we kept `static_argnames` instead? That seems clearer to me what arguments are static.

Dekermanjian commented:
Hi @jakevdp, anything I can improve upon with this?

Dekermanjian commented:
Hey @jakevdp, is there anything more that I can do here to push this over the finish line?

jakevdp commented:
Sorry for the slow response: all looks pretty good, except you have some undefined names in your code. Once that's fixed up and all the tests pass, we'll also need to squash all the changes into a single commit (see https://docs.jax.dev/en/latest/contributing.html#single-change-commits-and-pull-requests)

jakevdp commented on jax/_src/scipy/linalg.py:
```diff
@@ -2248,3 +2248,155 @@ def _binom(n, k):
+  else:
+    raise ValueError(f"Unrecognized method {method}. The two valid methods are either \"schur\" or \"eigen\".")
+
+  X = jnp.real(X) if jnp.isrealobj(C) else X
```
`jnp.isrealobj` is not available here after the recent build refactor. I'd rewrite it like this:
```python
if dtypes.issubdtype(C.dtype, np.complexfloating):
  X = X.real
```

jakevdp commented on jax/_src/scipy/linalg.py:
```diff
@@ -2248,3 +2248,155 @@ def _binom(n, k):
+
+  X = jnp.real(X) if jnp.isrealobj(C) else X
+
+  return jax.lax.cond(
```
Change this to `return lax.cond(` – `jax` is not imported in this namespace (to avoid circular imports)

Dekermanjian committed:
updated solve_sylvester method in accordance to recent build refactor

Dekermanjian commented:
> > Jake, do you know what the error AttributeError: module 'jax._src.numpy' has no attribute 'add' is referring to?
> 
> This is telling you that the `jax._src.numpy` submodule does not define the function `add`. Note that `jax.numpy`, which is exposed externally, and `jax._src.numpy`, which is used internally, are different modules despite the similar names. If you wish, you can update `jax._src.numpy` to import the `add` symbol here:
@jakevdp, I updated `solve_sylvester` to use broadcasting instead of calling `jnp.add.outer` and replaced `jnp.nan` with `float("nan")` because `jnp.nan` was not defined either. Locally, running `pytest -n auto tests/linalg_test.py -k test_ill_conditioned_sylvester` and `pytest -n auto tests/linalg_test.py -k test_solve_sylvester` and `pre-commit run --all` everything passes. Do you have a preference on re-exposing the `jnp.add` and `jnp.nan` methods as opposed to the changes I made?

jakevdp commented:
`jnp.nan` is an alias of `np.nan`, so if you'd like to use that here you can use numpy directly.

Dekermanjian committed:
Updated nan generation using numpy nan

jakevdp commented:
The doctest is failing:
```
??? >>> with jax.numpy.printoptions(precision=0):
??? ...   print(X)
Expected:
    [[ 1.e+00 -4.e-07]
      [-3.e-07  1.e+00]]
Got:
    [[ 1.e+00 -6.e-08]
     [ 3.e-08  1.e+00]]
```
You might need to add `suppress=True` to the printoptions. You can run the doctest locally with
```
pytest --tb=short --doctest-modules jax/_src/scipy/linalg.py
```

Dekermanjian committed:
updated solve_sylvester documentation example and bumped down tol for complex128 types in tests

Dekermanjian commented:
Thank you, @jakevdp. Being able to test issues locally is really helpful. I went ahead and updated the documentation example so that the doctests pass, and I had to bump down the tolerance for complex128 types form 1e16 to 1e12 for the tests to pass. I believe this may be an artifact of randomly generating the matrices in the test.

jakevdp commented:
Doctests are still failing, unfortunately:
```
Expected:
    [[1. 0.]
     [0.  1.]]
Got:
    [[ 1. -0.]
     [ 0.  1.]]
```
You could add `doctest: +SKIP` on the print statement if you don't want this line to be validated.

Dekermanjian committed:
skip doctest example validation

jakevdp commented on tests/linalg_test.py:
```diff
@@ -2115,6 +2122,69 @@ def testToeplitzConstructionWithKnownCases(self):
+    if jtu.test_device_matches(["gpu"]) and method == "schur":
+      self.skipTest("Schur not supported on GPU.")
+
+    tol = {np.float32: 3e-2, np.complex64: 3e-2, np.complex128: 1e-12}
```
It looks like there's still a failure at around the `1E-10` level.

Dekermanjian replied:
hmm, anyway I can test this locally in bulk? I ran this test at least 20 times manually and it was passing at `1e-12` level.

Dekermanjian replied:
@jakevdp, I updated the `tol` to `1E-9` level for `complex128` types. I ran the `complex128` test 1000 times and all randomly generated matrices passed at that `tol` level.

jakevdp replied:
I think if the solver is useful for complex types, we should test the regions in which it's useful. If the solver is not useful for complex types, we should raise an error when the user passes a complex type. If there's any gray area, we should test the regime that makes sense for testing, but also add a documentation note about any regimes where things may go wrong.

Dekermanjian replied:
Yes, I agree with you. It seems that the solver (similar to Scipy) has limitations when the size of the matrix increases. I am finding that when ether axis of the matrix increases beyond ~120 the solver fails to produce a good solution for complex128 types. I updated the tests to remain in that region of matrix sizes and I added a note on this degradation of results in the notes section of the docstring.

Dekermanjian committed:
updated complex128 tol in test_solve_sylvester

Dekermanjian committed:
updated tests to use smaller matrices. updated doctsring print options. added note on degradation of complex type resulst with increases to matrix size

Dekermanjian commented on jax/_src/scipy/linalg.py:
```diff
@@ -2250,3 +2250,156 @@ def _binom(n, k):
+    >>> B = jax.numpy.array([[5, 6], [7, 8]])
+    >>> C = jax.numpy.array([[6, 8], [10, 12]])
+    >>> X = jax.scipy.linalg.solve_sylvester(A, B, C)
+    >>> with jax.numpy.printoptions(precision=0, suppress=True): # doctest: +SKIP
```
Yes, I agree. I made the change in the latest push.

Dekermanjian commented:
Hey @jakevdp, why would the GPU produce different results compared to the CPU? Would this be specific to H100s or would you expect to see the same across all GPUs? Also, when testing on the GPU are we comparing the H100s results to CPU computed Scipy results?

jakevdp commented:
For this particular change, you're not touching any non-Python sources so you should be able to just use the most recent jaxlib release, assuming it's compatible with your GPU. Then you can run the Python parts of JAX directly from the repo. If you're compatible with cuda12 locally, then from the repository directory you can run
```
pip install -e .[cuda12]
```
One thing to try would be to make sure the GPU tests are run with `@jax.default_matmul_precision("float32")`, because the default on most hardware is much less precise.

Dekermanjian commented:
> For this particular change, you're not touching any non-Python sources so you should be able to just use the most recent jaxlib release, assuming it's compatible with your GPU. Then you can run the Python parts of JAX directly from the repo. If you're compatible with cuda12 locally, then from the repository directory you can run
> 
> ```
> pip install -e .[cuda12]
> ```
> 
> One thing to try would be to make sure the GPU tests are run with `@jax.default_matmul_precision("float32")`, because the default on most hardware is much less precise.

Thank you @jakevdp, is this supported on WSL2 or do I need to stick to Windows? I believe I have cuda13 installed is that supported?

jakevdp commented:
Platform support is summarized here: https://docs.jax.dev/en/latest/_autosummary/jax.default_matmul_precision.html Windows has very spotty support, especially for accelerators.

jakevdp commented:
Another possibility is just skipping the problematic tests on GPU with a `# TODO(jakevdp)` comment to re-enable them.

Dekermanjian committed:
bumped tol on float32 anc complex64 added decorator to ensure gpu default precision is float32

Dekermanjian commented:
Hey @jakevdp, I was able to set up a dev environment on WSL2 thanks to your help earlier. I bumped the tolerances for float32 and complex64 types and added `@jax.default_matmul_precision("float32")` to the test which was key for passing tests on GPU. I ran tests at a thousand iterations locally and they were passing crossing my fingers they pass the CI testing.
